### ISSUE TITLE
Incorrectly reduces if statements, changing function behavior.

### ISSUE THREAD

**Issue by mythmon** [MaiaVictor/LJSON#13]
The partial evaluator seems to be a bit too greedy when evaluating if statements, and can change the  behavior of the code

``` javascript
function f(x, y) {
  if (x > y) {
    return x;
  } else {
    return y;
  }
}

f(1, 2);  // returns 2
f(2, 1);  // returns 2

var enc = LJSON.stringify(f);  // returns '(v0,v1)=>(v1)'
var f_ = LJSON.parse(s);
f_(1, 2);  // returns 2
f_(2, 1);  // returns 1. ERROR
```

**Comment by VictorTaelin** [comment]
LJSON extends JSON with nothing but pure functions and applications - nothing more. Both `if` and `>` are out of the specs. If you want them, you need an environment of primitive functions which include those. It would be interesting to include them on the standard lib, which currently is all but empty. Something like:

```
// add to `withStdLib` on LJSON.js
"if" : function(cond, case_true, case_false){ return cond ? case_true : case_false; }
"<" : function(a,b){ return a<b; }
```

Then you can write:

```
function f($, x, y){
    return $("if", $("<", a, b), a, b);
};
var enc = LJSON.stringify(f);
var f_ = LJSON.parseWithStdLib(enc); // notice 
```

This is the worst thing about LJSON IMO, it would be great to have a LJSON-like lib extended with the set of native JS operations and operators, but I believe that would be impossible. At least, I can't think of a way to make "stringify" work when you have native operators.

**Comment by mythmon** [comment]
I'd assume LJSON would aim to only serialize/deserialize functions correctly (ie, without changing behavior). If that is the case, it should refuse to serialize that function, instead of returning a different function. Is that assumption wrong?

**Comment by VictorTaelin** [comment]
Yes, it aims to serialize/deserialize **functions** correctly. That includes function definition (i.e., `function(vars...){return body}`) and application (`fun(arg0, arg2, arg3...)`), AKA substitution. The issue here is that `if` and `>` are not functions, nor function applications - those are **specific JavaScript operations**. LJSON doesn't know how to deal with them. It can't serialize **arbitrary JavaScript programs**. It would be awesome if it did, but I don't think that is even possible.

The purpose of LJSON is to pass code around safely. It guarantees that a serialized/deserialized is function 100% safe. It can not do anything that you don't allow, such as sending TCP messages. With LJSON, you are able to receive code from your clients and just execute it, no need for sandboxes, no safety worries. If LJSON allowed you to serialize arbitrary JS programs, that safety guarantee would be gone. What if a client included something like `install_virus()` in the middle of his code? There's no way to tell. You could write a whole game engine in LJSON. But, for that, you need to program all of it specifically on it, not in JavaScript.

**Comment by mythmon** [comment]
I understand that the function I gave as an example does not meet the definitions of purity that LJSON can deal with. I'm fine with that. I'm not expecting LJSON to magically purify the function. What I expect is that LJSON will recognize this and throw an error.

**Comment by VictorTaelin** [comment]
Ah, I see. That would be very nice, but I also don't think that is possible. At least, I can't think in any way of implementing that. LJSON's stringify algorithm assumes there are only functions and function applications. It works by applying argument-collecting-closures to functions. If those closures are used with JS operations (say, `a+b` instead of `a(b)`), JavaScript will just do its type-coercion thing on them, turning them into strings and the whole thing fails silently. I don't think there is a way to detect that those argument-collecting-closures were used as arguments of `+`, `if` and cia. So, in short, trying to stringify functions with native JS ops is undefined behavior and I don't think it can be better... sadly.
